draft-secure-credential-transfer.md

```diff
@@ -207,10 +207,10 @@ Sender device may terminate the secure credential transfer by deleting the mailb
 
 The Provisioning Information is the data transfered via the Relay Server between the Sender device and Receiver device. Each use case defines its own specalized Provisioning Information format, but all formats must at least adhear to the following structure. Formats are free to define new top level keys, so clients shouldn't be surprised if a message of an unexpected format has specalized top level keys.
 
-| Key                | Type       | Description 
-| ------             | ---        | --- 
-| format             | String     | The Provisioning Information format that the message follows. This is used by the Sender device and Receiver device to know how to parse the message.
-| genericSharingData | Dictionary | A dictionary of generic sharing data that can be used for cross platform credential transfers. See each format's specification for exact fields.
+| Key           | Type       | Description 
+| ------        | ---        | --- 
+| format        | String     | The Provisioning Information format that the message follows. This is used by the Sender device and Receiver device to know how to parse the message.
+| genericData   | Dictionary | A dictionary of generic sharing data that can be used for cross platform credential transfers. See each format's specification for exact fields.
```

Comment:
Wanted to circle back on a topic we discussed earlier and get further thoughts. 

Context: 
For hotel keys, the group proposed using tope level proprietary OEM fields to embed provisioning material.

My thoughts:
We are trying to establish a standard of using `format` and `sharingData` as a way to differentiate different credential provisioning structures. However, with hotel keys, it feels awkward that we would suddenly break this pattern and have an empty genericData. From the argument I'm hearing, it's because hotel keys requires apple and android specific fields and we already have similar top level OEM containers. While I can understand the convenience of this approach, I think it's too much of a happenstance. Also, while not enforceable, it breaks the principle of having the top level oem data containers be proprietary. In the hotel case, both containers would need to adhere to a strict standard. 

I would advocate putting all provisioning material into the `genericData` as that is the only field the `format` is currently coupled with. Otherwise, I think it weaken the standard and sets a precedent for coupling `format` to additional fields on the top level.

Let me know if this discussion will be better served over email. Thanks!

Reply:
> In the hotel case, both containers would need to adhere to a strict standard.

In the hotel case it will adhere to a standard, as you mentioned, it is a required for the system to work. We just haven't written / formalized that standard. Even though for cross platform hotel sharing all device OEM provisioning information will need to be included in the invite each OEM will only parse / read their own Information. The sender will simply be a pass through of the OEM data from the partner to the recipient.

> I would advocate putting all provisioning material into the genericData as that is the only field the format is currently coupled with. Otherwise, I think it weaken the standard and sets a precedent for coupling format to additional fields on the top level.

This is why I'm not a fan of the including `genericData` in this spec.

If we think having format defined top level keys will weaken the standard what do you think about having `format` and `content` (name WIP) as the only two top level keys. And then all credential transfer information would live inside content. i.e.:

```
{
	"format" : "digitalwallet.carkey.ccc"
	"content": {
	    "genericSharingData": {
	        // ...
	    },
	    "appleSharingData": { ... },
	    "androidSharingData": { ... },
	}
}
```

Then if we wanted to add new top level keys to the IETF standard there would be no risk that those keys collide with the keys used in the various format specs. For example, we could add a `messageIdentifier` field to the IETF spec.

Reply:
> In the hotel case it will adhere to a standard, as you mentioned, it is a required for the system to work.

> Even though for cross platform hotel sharing all device OEM provisioning information will need to be included in the invite each OEM will only parse / read their own Information.

I think my reservation here is mixing a container for proprietary use with provisioning information. Forcing a standardization on a container meant for proprietary use kind of defeats its nature.

> If we think having format defined top level keys will weaken the standard what do you think about having format and content (name WIP) as the only two top level keys. And then all credential transfer information would live inside content. i.e.:

I like the idea of having a content and pending other's thoughts, I'm good with this.  Let me also provide an example and last argument on how it would work with current structure: 

```
{ 
        "format" : "digitalwallet.carkey.ccc", 
        "content": { // this structure is fully governed by the format type and its external source
        ​     // structure here would be outlined in the CCC. Contains sharingData, sharingDataType, etc.
       ​ },
       ​ "appleSharingData" : { ... }, // proprietary use, structure not defined but key may be defined in IETF to avoid collision
       ​ "androidSharingData" : { ... }, 
       ​ "xxxSharingData" : { ... },
}

{ 
        "format" : "digitalwallet.hotelkey.hyatt",
        "content": { 
        ​    "appleHotelData": { ... }, 
        ​    "androidHotelData": { ... }, 
       ​ }, 
       ​ ​"appleSharingData" : { ... },
       ​ "androidSharingData" : { ... }, 
       ​ "xxxSharingData" : { ... }
}
```

I think the structure of content(or genericData) need not be defined via IETF. If it contains apple/android specific containers, that's up to the discretion of that credential standard. IETF need not try to optimize by sharing existing proprietary containers. I believe this also helps simplify onboarding new credentials/formats as it removes the guess work of trying to optimize with fields outside its domain.

Reply:
> I think my reservation here is mixing a container for proprietary use with provisioning information. Forcing a standardization on a container meant for proprietary use kind of defeats its nature.

I agree with this, although I don't feel very strongly about it. I see the benefit in separating true proprietary or secret sauce keys from device OEM opaque blobs; where even though the contents of the blob is opaque all device OEMs know that the blob is effectively a bearer token.

Since we are settled on "format" + "content" I think this discussion doesn't need to block this PR. I'll added this discussion to the agenda next week.

---

I was thinking it would work like:

```
{
   "format": "digitalwallet.hotelkey.hyatt",
   "content": {
      "appleHotelData": {  ... },
      "androidHotelData": {  ... },
      // Device OEM Proprietary Containers
      "appleSharingData": {  ... },
      "androidSharingData": {  ... },
      "xxxSharingData": {  ... }
   }
}
```

That way there is no risk of future EITF spec changes introducing new top level keys that collide with proprietary fields. But since the usage of "appleSharingData" / "androidSharingData" is up to each device OEM the data could be put anywhere.